woocommerce/woocommerce
# k6 perf tests: Fix parsing of COT/HPOS environment variable

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





#### Problem

If you want to run our k6 tests on a site with HPOS is enabled, you need to pass an environment variable called `COT` with value `true`. But remember that k6 receives environment variables as strings. It will recognize the `COT` environment variable as a stringified `"true"` instead of a boolean `true`.

<img width="866" alt="Screenshot 2023-10-22 at 9 35 23 PM" src="https://github.com/woocommerce/woocommerce/assets/4509348/74ba385f-1411-48d8-9bf1-be4931eb798e">

And it turns out that our k6 tests check for a boolean `true` instead of a stringified `"true"` as you can see here:

<img width="895" alt="Screenshot 2023-10-22 at 9 39 32 PM" src="https://github.com/woocommerce/woocommerce/assets/4509348/9a98fb1d-827b-4e2e-a669-94c057bfa7a1">

Therefore, the tests would run as if HPOS was disabled even though we've properly set the `COT` variable to `true`. This is the reason why I got around 500+ failed checks when I ran our k6 tests against our performance test site with `8.3.0.20` even after setting `COT` to `true`.

<img width="768" alt="cot-true site-hpos-on" src="https://github.com/woocommerce/woocommerce/assets/4509348/1092fc1d-a324-4322-9c8b-073641174f61">

#### Solution

This PR solves the problem by ensuring that the variable `cot_status`, which the PR renames to `hpos_enabled`, will receive a boolean value.

The PR also takes the opportunity to:

- update all `COT` references to `HPOS` in our k6 test directory, including the README
- change the `COT` environment variable to `HPOS`, which defaults to `true`.
- fix [one assertion](https://github.com/woocommerce/woocommerce/commit/4e2702e514ce88a4d71862ce8dc79a15b3cf2d90)
- also prevent the other boolean variable from getting a string value [here](https://github.com/woocommerce/woocommerce/commit/00e5b6263d4a9084f822dad3f6e2dfd31dcc43b4). It should still default to `false`, as originally intended.
- Our `pnpm env:dev` setup script launches the wp-env environment with HPOS disabled 🤷‍♂️ . It only enables after logging in to wp-admin through the UI. We have observed a similar behavior in the E2E tests. To handle this, at least for the meantime, this PR explicitly enables HPOS in the `plugins/woocommerce/tests/performance/bin/init-sample-products.sh` file as done in https://github.com/woocommerce/woocommerce/pull/40930/commits/23cb4af4874de7106c3e39d6a4c2a7a6378eeaf1.

I'm also looping in @rrennick as a reviewer because he also did a lot of fixes in the k6 perf tests.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Checkout this branch
1. Install k6 in your local machine.
1. Log in to the nightly performance test site (get details from the secret store). And make sure that:
   - HPOS is enabled there
   - Latest nightly version of WooCommerce is installed
1. Run the k6 tests against our nightly performance test site.

   ```bash
   cd plugins/woocommerce/tests/performance

   k6 run \
      -e URL="<...>" \
      -e HOST="<...>" \
      -e A_USER="<your admin user in that site>" \
      -e A_PW="<your admin pw>" \
      -e C_USER="<your customer username>" \
      -e C_PW="<customer password>" \
      -e C_UID="<customer user id>" \
      -e P_ID=22733 \
      -e HPOS=true \
      -e A_ACC_LOGIN=true \
      tests/wc-regression-test-load.js
   ```

   Note that `A_ACC_LOGIN` should be true. Otherwise, there would be a lot of failures. I think it deserves to be looked at, but is already outside the scope of this PR.

1. In the report, verify that you only get a few errors (less than a hundred).
    <img width="781" alt="cot-false site-hpos-off" src="https://github.com/woocommerce/woocommerce/assets/4509348/cee9de54-5f04-4f8d-9bf6-00a169433107">
1. Rerun the test without the HPOS environment variable. The report should still show very few errors.
1. Now, manually disable HPOS in the perf test site.
1. Rerun the test but with HPOS set to false:

   ```bash
   cd plugins/woocommerce/tests/performance

   k6 run \
      -e URL="<...>" \
      -e HOST="<...>" \
      -e A_USER="<your admin user in that site>" \
      -e A_PW="<your admin pw>" \
      -e C_USER="<your customer username>" \
      -e C_PW="<customer password>" \
      -e C_UID="<customer user id>" \
      -e P_ID=22733 \
      -e HPOS=false \
      -e A_ACC_LOGIN=true \
      tests/wc-regression-test-load.js
   ```

1. The report should still show very few errors.
2. Turn HPOS in the perf test site back on.
1. Make sure the k6 tests in this PR passed.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
